### PR TITLE
Mc issue 318 permissions refactor

### PR DESCRIPTION
**Pull request by maxachis**
#### Fixes

* Partially addresses https://github.com/Police-Data-Accessibility-Project/data-sources-app/issues/162 
* Fully addresses https://github.com/Police-Data-Accessibility-Project/data-sources-app/issues/387
* Fully addresses https://github.com/Police-Data-Accessibility-Project/data-sources-app/issues/322

#### Description

* Replace `@api_key` with `@api_key_required` decorator.
* Replace `Bearer` with `Basic` authentication
* Add `/auth/permissions` endpoint and associated logic/tests
* Implement new logic for `@api_key_required`. Update endpoints, related logic, and tests
* Add required permissions for some endpoints. Updated tests accordingly.
* Remove role logic
* Remove Search Tokens and Access Tokens logic

#### Testing

* Run all automated tests from `tests` directory.
* Additionally, access Swagger UI documentation and confirm proper documentation and functionality of new endpoints.

#### Performance

* New permissions logic is roughly comparable in complexity to previous api_key logic, so performance impact is mixed.
* Some tests added but other tests removed, so test overhead impact is mixed
* Removal of `search_tokens` and `access_tokens` tables and addition of `permissions` and `user_permissions` table has mixed impact on storage overhead of database.

#### Docs

* Swagger documentation updated. 

#### Breaking Changes

* The majority of endpoints (absent ones such as /login and /api_key, which utilize email and password, and performance-dependent endpoints such as /typeahead-suggestions) now require either Bearer or Basic authorization.
* Some endpoints which previously used Bearer <API_KEY> authorization now use Basic <API_KEY> authorization.
* For a limited set of endpoints, the authorization is now Bearer <JWT>, with the JWT being provided by one of the /login endpoints.
* For these endpoints, only select users will have permission. To obtain a test user with elevated permissions, a developer will need to make POST request to the /dev/create-test-user-with-elevated-permissions endpoint, providing the `dev_password`, , which must match a server-set `DEVELOPMENT_PASSWORD`
* On the development side, most resources and their associated tests have been modified to accommodate the changes to permissions.

### CODE REVIEW

**Comment by josh-chamberlain** [comment]
@maxachis which bits of 318 does this address? Is it more accurate to say it addresses https://github.com/Police-Data-Accessibility-Project/data-sources-app/issues/162?

File: README.md [hunk 1]
```diff
@@ -65,7 +65,7 @@ The environment variables are as follows:
 * DEV_DB_CONN_STRING: Used to connect to the dev database
 * VITE_VUE_API_BASE_URL: The base URL for the API
 * VITE_VUE_APP_BASE_URL: The base URL for the UI
-* SECRET_KEY: Used to sign cookies
```
Comment (josh-chamberlain) [inline README.md:68]: on my end, the SECRET_KEY looks like an API key that doesn't seem to refresh. Is that accurate? Were the two concepts somehow mixed?

Comment (maxachis) [reply README.md:69]: Technically, a SECRET_KEY can be anything. It's used as a key for the encryption and decryption of values (in this case, to ensure cookies aren't being forged). I did rename the environment variable to `FLASK_APP_SECRET_KEY` to be more precise. The `JWT_SECRET_KEY` servers a similar function, but in this case to encrypt the JWTs.

Comment (josh-chamberlain) [reply README.md:69]: I see. The root of my question: is an API key no longer required to run the app locally? It looks like the envars bit of the README is still a bit out of date

Comment (josh-chamberlain) [reply README.md:69]: nice, these instructions are helpful. Agreed that the swagger docs should tell people _how_ the `...test-user...` endpoint works, but we need to tell people _that they need to use it._

File: .github/workflows/run_pytest.yml [hunk 2]
```diff
@@ -6,7 +6,7 @@ on:
 jobs:
   run_tests:
     env:
-      SECRET_KEY: ${{ secrets.SECRET_KEY }}
+      FLASK_APP_SECRET_KEY: ${{ secrets.SECRET_KEY }}
```
Comment (josh-chamberlain) [inline .github/workflows/run_pytest.yml:9]: does this instance need to change to `FLASK_APP_COOKIE_ENCRYPTION_KEY`? What about other instances of `SECRET_KEY` and `FLASK_APP_SECRET_KEY` in the code?

File: CONTRIBUTING.md [hunk 3]
```diff
@@ -2,6 +2,21 @@
...
+`Basic` endpoints are endpoints which any registered user can utilize and which do not require elevated permissions.
+
+`Bearer` endpoints are endpoints which require elevated permissions.
+
+To create a user with elevated permissions, use the `/dev/create-test-user-with-elevated-permissions` endpoint, and follow instructions provided. 
+That endpoint will provide you with a randomly generated email, password, and api key to use to test endpoints. The permissions are such that all endpoints should be available to you for use.
+To obtain a short-lived JWT for use in the `Bearer` endpoints, use the `/login` endpoint with the provided email and password to receive the requisite JWT.
+
+
 ## Installation
```
Comment (josh-chamberlain) [inline CONTRIBUTING.md:20]: could we keep this at the top of the file, please?

File: .github/workflows/run_pytest.yml (hunk 2 above)
Comment (maxachis) [reply .github/workflows/run_pytest.yml]: Yes it does, thank you for catching that! Other instances have already been changed!

File: CONTRIBUTING.md (hunk 3 above)
Comment (maxachis) [reply CONTRIBUTING.md]: Done!

**Review by joshuagraber — APPROVED** [review]

**Pull request merged**